2021/01/29/intro-to-typescript.html.md

```diff
@@ -0,0 +1,287 @@
+
+```typescript
+type boolOrString = boolean | string;
+
+let myVar: boolOrString;
+
+myVar = true;
+myVar = 'hello!'
+```
+
+This is 100% valid. myVar can hold either a boolean OR a string value. However typescript will force you to disambiguate which type it currently holds before you can do anything unsafe with it. For example if we have a function that takes a union type
+
+```typescript
+function upperBool(arg: boolean | string): string {
+    if (typeof arg == 'string') {
+        return arg.toUpperCase()
+    }
+    else if (typeof arg == 'boolean') {
+        return arg.toString().toUpperCase()
+    }
+}
+```
+
+It looks like our function could fall through and return undefined, and
+theoretically in JavaScript it could, but any such usage would trigger an
+error in TypeScript. TypeScript "knows" that we've handled both possible
+types of arg and so the if/else definately does not fall through.
+
+## Intersection Types
+
+An intersection type combines two or more interfaces into a single compound interface.
+
+```typescript
+interface This {
+    aProperty: string;
+}
+
+interface That {
+    bProperty: boolean;
+}
+
+type Both = This & That;
+
+const both: Both = {
+    aProperty: 'funny',
+    bProperty: true
+}
+```
+
+## Classes
+
+I won't get deep into TS classes, but I want to mention them. TS has classes.
+They can transpile to ES6 classes or ES2015 prototypal code. They are very
+powerful. They are, however, optional, as classes are still somewhat "new" to
+many JS devs.
+
+## Enums
+
+TypeScript supports enum declarations. An enum is a type that can only hold
+one of a specific set of possible values.
+
+## Generics
+
+Generics are a mechanism for declaring "complex" types like arrays and
+promises. Creating new generics is somewhat advanced, but a basic
+understanding is required to use TS effectively.
+
+```typescript
+const arrayOfStrings: string[];
+const anotherArrayOfStrings: Array<string>;
+```
+
+Both of these syntaxes are equivalent. They both declare "an array of strings".
+
+```typescript
+let promiseMeAString: Promise<string>
+```
+
+This declares a promise that is expected to resolve to a string type value.
+
+## Indexible Types
+
+Indexible types allow the developer to specify the key and value types
+expected when using the square-bracket operator to access object or array
+properties.
+
+The syntax is a little funny in that it includes a "placeholder" field that
+TS ignores and can be set to anything.
+
+```typescript
+interface stringKeysBoolVals {
+    [x: string]: boolean
+}
+```
+
+`x` is the placeholder. You could just as well put in `doTheFunkyChicken`;
+it's syntactically required, but the value is ignored. I mention this
+specifically because it tripped me up at first.
+
+This is valid usage
+
+```typescript
+const myCollection: stringKeysBoolVals = {
+    "value1": true,
+    "value2": false
+} 
+```
+
+Invalid usage
+
+```typescript
+const myCollection: stringKeysBoolVals = {
+    0: "foobar"
```

Comment:
I'm a bit thrown off by the the "Invalid usage" part of this example because the body of it only has one line vs the two lines of the body in the "valid usage" part, but I think there are a couple of other aspects of the example that I think you are illustrating implicitly here that you might want to remark on as well. (Let's review this together.)